open_instruct/utils.py

```diff
@@ -38,6 +41,286 @@
+
+    raw_datasets = DatasetDict()
+    raw_train_datasets = []
+    raw_val_datasets = []
+    frac_or_sample_list = []
+    for (ds, frac_or_samples), ds_config in zip(dataset_mixer.items(), configs):
+        frac_or_sample_list.append(frac_or_samples)
+        for split in splits:
+            # if dataset ends with .json or .jsonl, load from file
+            if ds.endswith(".json") or ds.endswith(".jsonl"):
+                dataset = load_dataset("json", data_files=ds, split=split)
+            else:
+                try:
+                    # Try first if dataset on a Hub repo
+                    dataset = load_dataset(ds, ds_config, split=split)
+                except DatasetGenerationError:
+                    # If not, check local dataset
+                    dataset = load_from_disk(os.path.join(ds, split))
+
+            # shuffle dataset if set
+            if shuffle:
+                dataset = dataset.shuffle(seed=42)
+
+            # assert that needed columns are present
+            if need_columns:
+                if not all(col in dataset.column_names for col in need_columns):
+                    raise ValueError(f"Needed column {need_columns} not found in dataset {dataset.coulmn_names}.")
+
+            # handle per-case conversions
+            # if "instruction" and "output" columns are present and "messages" is not, convert to messages
+            if (
+                "instruction" in dataset.column_names
+                and "output" in dataset.column_names
+                and "messages" not in dataset.column_names
+            ):
+                dataset = dataset.map(instruction_output_to_messages, num_proc=10)
+            elif (
+                "prompt" in dataset.column_names
+                and "completion" in dataset.column_names
+                and "messages" not in dataset.column_names
+            ):
+                dataset = dataset.map(prompt_completion_to_messages, num_proc=10)
+            elif "conversations" in dataset.column_names and "messages" not in dataset.column_names:
+                dataset = dataset.map(conversations_to_messages, num_proc=10)
+            elif (
+                "question" in dataset.column_names
+                and "response" in dataset.column_names
+                and "messages" not in dataset.column_names
+            ):
+                dataset = dataset.map(question_response_to_messages, num_proc=10)
+            elif (
+                "query" in dataset.column_names
+                and "answer" in dataset.column_names
+                and "messages" not in dataset.column_names
+            ):
+                dataset = dataset.map(query_answer_to_messages, num_proc=10)
+            elif (
+                "query" in dataset.column_names
+                and "response" in dataset.column_names
+                and "messages" not in dataset.column_names
+            ):
+                dataset = dataset.map(query_response_to_messages, num_proc=10)
```

Comment:
Probably raise an unsupported error if none is matched?

Reply:
Most of the time it'll error if it doesn't match.